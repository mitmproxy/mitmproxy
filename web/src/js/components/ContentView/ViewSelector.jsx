import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { setContentView } from '../../ducks/ui/flow';
import Dropdown from '../common/Dropdown'


ViewSelector.propTypes = {
    contentViews: PropTypes.array.isRequired,
    activeView: PropTypes.string.isRequired,
    setContentView: PropTypes.func.isRequired
}

export function ViewSelector ({contentViews, activeView, setContentView}){
    let inner = <span> <b>View:</b> {activeView.toLowerCase()} <span className="caret"/> </span>

    return (
        <Dropdown dropup className="pull-left" btnClass="btn btn-default btn-xs" text={inner}>
            {contentViews.map(name =>
                <a href="#" key={name}  onClick={e => {e.preventDefault(); setContentView(name)}}>
                    {name.toLowerCase().replace('_', ' ')}
                </a>
                )
            }
        </Dropdown>
    )
}

export default connect (
    state => ({
        contentViews: state.settings.contentViews || [],
        activeView: state.ui.flow.contentView,
    }), {
        setContentView,
    }
)(ViewSelector)
