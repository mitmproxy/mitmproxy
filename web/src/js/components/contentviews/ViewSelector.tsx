import React from "react";
import {useAppSelector} from "../../ducks";
import Dropdown, {MenuItem} from "../common/Dropdown";

type ViewSelectorProps = {
    value: string
    onChange: (string) => void
}

export default function ViewSelector({value, onChange}: ViewSelectorProps) {
    const contentViews = useAppSelector(state => state.backendState.contentViews || []);

    let inner = <span><i className="fa fa-fw fa-files-o"/>&nbsp;<b>View:</b> {value.toLowerCase()} <span
        className="caret"/></span>

    return (
        <Dropdown
            text={inner}
            className="btn btn-default btn-xs"
            options={{placement: "top-start"}}>
            {contentViews.map(name =>
                <MenuItem key={name} onClick={() => onChange(name)}>
                    {name.toLowerCase().replace('_', ' ')}
                </MenuItem>
            )}
        </Dropdown>
    )
}
