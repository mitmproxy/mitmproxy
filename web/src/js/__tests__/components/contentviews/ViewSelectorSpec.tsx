import * as React from "react";
import ViewSelector from "../../../components/contentviews/ViewSelector";
import { fireEvent, render, screen } from "../../test-utils";

test("ViewSelector", async () => {
    const onChange = jest.fn();
    const { asFragment } = render(
        <ViewSelector value="Auto" onChange={onChange} />,
    );
    expect(asFragment()).toMatchSnapshot();
    fireEvent.click(screen.getByText("auto"));
    expect(asFragment()).toMatchSnapshot();
    fireEvent.click(screen.getByText("raw"));
    expect(onChange).toBeCalledWith("Raw");
});
