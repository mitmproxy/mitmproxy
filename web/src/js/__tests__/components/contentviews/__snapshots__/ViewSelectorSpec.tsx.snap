// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewSelector 1`] = `
<DocumentFragment>
  <a
    class="btn btn-default btn-xs"
    href="#"
  >
    <span>
      <i
        class="fa fa-fw fa-files-o"
      />
       
      <b>
        View:
      </b>
       auto 
      <span
        class="caret"
      />
    </span>
  </a>
</DocumentFragment>
`;

exports[`ViewSelector 2`] = `
<DocumentFragment>
  <a
    class="btn btn-default btn-xs open"
    href="#"
  >
    <span>
      <i
        class="fa fa-fw fa-files-o"
      />
       
      <b>
        View:
      </b>
       auto 
      <span
        class="caret"
      />
    </span>
  </a>
  <ul
    class="dropdown-menu show"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <li>
      <a
        href="#"
      >
        auto
      </a>
    </li>
    <li>
      <a
        href="#"
      >
        raw
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
